src/tasks/wifi.c

```diff
@@ -491,10 +518,104 @@ static void do_beacon()
         };
         send_beacon(serial, ips);
 
-        /* Now flush all incomming data since we don't care about it. */
+
+        /* Now flush all incoming data since we don't care about it. */
         serial_purge_rx_queue(serial);
+
+do_beacon_exit:
+        xSemaphoreGive(state.connection_mutex);
+}
+
+static bool start = false;
+
+
+/* TODO this should get broken out to a modular design */
+
+/* Camera control for Hero 4+, including session */
+/*
+static void send_camera_control(struct Serial* serial, const char* ips[])
+{
+    if (start) {
+            serial_write_s(serial, "GET /gp/gpControl/command/shutter?p=1 HTTP/1.0\r\n\r\n");
+    }
+    else {
+            serial_write_s(serial, "GET /gp/gpControl/command/shutter?p=0 HTTP/1.0\r\n\r\n");
+    }
+    start = !start;
+}
+*/
+
+/* Camera control for Hero 2,3
+ * Format is: GET /bacpac/SH?t=<wifi-password>&p=<shutter-param>
+ */
+static void send_camera_control(struct Serial* serial)
+{
+    const char preamble[] = "GET /bacpac/SH?t=";
+    const char epilogue[] = " HTTP/1.0\r\n\r\n";
+    const char shutter_param[] = "&p=";
+    const int command_size = sizeof(preamble) + WIFI_PASSWD_MAX_LEN + sizeof(shutter_param) + 2 + sizeof(epilogue) + 5;
+    char command[command_size];
+
+    strncpy(command, preamble, sizeof(preamble));
+    strncat(command, getWorkingLoggerConfig()->ConnectivityConfigs.wifi.client.passwd, WIFI_PASSWD_MAX_LEN);
+    strncat(command, shutter_param, sizeof(shutter_param));
+    strncat(command, start ? "%01" : "%00", 3);
+    strncat(command, epilogue, sizeof(epilogue));
+
+    pr_info_str_msg(LOG_PFX "Camera control: ", start ? "starting" : "stopping");
+
+    serial_write_s(serial, command);
+    start = !start;
 }
 
+static void do_camera_control()
+{
+        const struct esp8266_ipv4_info* client_ipv4 = get_client_ipv4_info();
+
+        if (! is_valid_ipv4(client_ipv4->address))
```

Comment:
Check indentation.  Good use of rapid return.  Put bang next to method call.

Reply:
done